algorithms/src/snark/marlin/ahp/prover/round_functions/third.rs

```diff
@@ -21,249 +20,286 @@ use crate::{
         fft_precomputation: &FFTPrecomputation<F>,
         ifft_precomputation: &IFFTPrecomputation<F>,
-    ) -> (Sum<F>, Lhs<F>, Gpoly<F>) {
-        let mut job_pool = snarkvm_utilities::ExecutionPool::with_capacity(2);
-        job_pool.add_job(|| {
-            let a_poly_time = start_timer!(|| format!("Computing a poly for {label}"));
-            let a_poly = {
-                let coeffs = cfg_iter!(arithmetization.val.as_dense().unwrap().coeffs())
-                    .map(|a| v_H_i_alpha_v_H_i_beta * a)
-                    .collect();
-                DensePolynomial::from_coefficients_vec(coeffs)
-            };
-            end_timer!(a_poly_time);
-            a_poly
-        });
-
-        let (row_on_K, col_on_K, row_col_on_K) =
-            (&arithmetization.evals_on_K.row, &arithmetization.evals_on_K.col, &arithmetization.evals_on_K.row_col);
-
-        job_pool.add_job(|| {
-            let b_poly_time = start_timer!(|| format!("Computing b poly for {label}"));
-            let alpha_beta = alpha * beta;
-            let b_poly = {
-                let evals: Vec<F> = cfg_iter!(row_on_K.evaluations)
-                    .zip_eq(&col_on_K.evaluations)
-                    .zip_eq(&row_col_on_K.evaluations)
-                    .map(|((r, c), r_c)| alpha_beta - alpha * r - beta * c + r_c)
-                    .collect();
-                EvaluationsOnDomain::from_vec_and_domain(evals, non_zero_domain)
-                    .interpolate_with_pc(ifft_precomputation)
-            };
-            end_timer!(b_poly_time);
-            b_poly
-        });
-        let [a_poly, b_poly]: [_; 2] = job_pool.execute_all().try_into().unwrap();
-
-        let f_evals_time = start_timer!(|| format!("Computing f evals on K for {label}"));
-        let mut inverses: Vec<_> = cfg_iter!(row_on_K.evaluations)
-            .zip_eq(&col_on_K.evaluations)
-            .map(|(r, c)| (beta - r) * (alpha - c))
+        assignment: DensePolynomial<F>,
+        arith: &MatrixArithmetization<F>,
+        c_i_to_k: BTreeMap<usize, Vec<usize>>,
+        alpha: F,
+        combiner: F,
+    ) -> Result<LinevalInstance<F>, anyhow::Error> {
+        let sumcheck_time = start_timer!(|| format!("Compute LHS of sumcheck for {_label}"));
+
+        let v_R: SparsePolynomial<F> = constraint_domain.vanishing_polynomial();
+
+        let m_at_alpha_evals_time = start_timer!(|| format!("Compute m_at_alpha_evals parallel for {_label}"));
+        let mut m_at_alpha_evals: BTreeMap<usize, F> = cfg_iter!(c_i_to_k)
+            .map(|(&c_i, ks)| {
+                let mut m_at_c_i = F::zero();
+                for &k in ks {
+                    let col_at_k = arith.evals_on_K.col[k];
+                    let row_at_k = arith.evals_on_K.row[k];
+                    let row_col_val_at_k = arith.evals_on_K.row_col_val[k];
+                    let l_row_at_k = constraint_domain.size_inv * v_R.evaluate(alpha) / (alpha - row_at_k);
```

Comment:
We can evaluate `v_R` at alpha once outside the map

Reply:
Ditto for the inversion: we should use batch_inversion to compute `(alpha - row_at_k).inverse()`. Otherwise this loop will be way too slow.

Reply:
Great points, will be mindful of inversion in the future - c72f41937